### ISSUE TITLE
Relaxed type checking in multiple places.

### ISSUE THREAD

**Issue by trocher** [vyperlang/vyper#3232]
### Version Information

* vyper Version (output of `vyper --version`): 0.3.8+commit.4ae2527f
* OS: OSX
* Python Version (output of `python --version`): 3.8.0

### What's your issue about?

Note that this bug is not present in v0.3.7 but only in the latest commits of the master branch.

The restriction in the following cases have been relaxed:
- the type of Events' fields. 
- the type of Structs' fields.
- the type of `VariableDecl`.
- the type of HashMaps' values.
- the type of functions' arguments.
- the return type of functions.

Note that this list is probably not exhaustive.

In certain cases, using types that were forbidden in v0.3.7 lead the contract to compile when in other case, the compiler is  panicking. 

My guess is that the issue is caused by `type_from_annotation` no longer taking a location as argument. Hence I would assume type checking has been relaxed in other places as well.

The following examples compiles (where they would not in `v0.3.7`):
```Vyper
event A:
    a:uint256
x:A
```
```Vyper
event A:
    a:uint256
x: HashMap[uint256,A]
```
```Vyper
event A:
    a:uint256
struct B:
    a:A
```
The following examples lead the compiler to panic with the corresponding outputs

```Vyper
event Foo:
    a:HashMap[uint256, uint256]
```
```Vyper
b:HashMap[uint256, uint256] 

@external
def bar():
    a:HashMap[uint256, uint256] = self.b
```
```Vyper
x: HashMap[uint256, uint256]
@external
def foo() -> HashMap[uint256, uint256]:
    return self.x
```
```Bash
Error compiling: tests/customs/code.vy
Traceback (most recent call last):
  File "/Users/trocher/Documents/thesis/vyper/venv/bin/vyper", line 11, in <module>
    load_entry_point('vyper==0.3.8', 'console_scripts', 'vyper')()
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 57, in _parse_cli_args
    return _parse_args(sys.argv[1:])
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 154, in _parse_args
    compiled = compile_files(
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 294, in compile_files
    compiler_data = vyper.compile_codes(
  File "/Users/trocher/Documents/thesis/vyper/vyper/evm/opcodes.py", line 226, in _wrapper
    return fn(*args, **kwargs)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/__init__.py", line 141, in compile_codes
    exc_handler(contract_name, exc)
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 189, in exc_handler
    raise exception
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/__init__.py", line 138, in compile_codes
    out[contract_name][output_format] = OUTPUT_FORMATS[output_format](compiler_data)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/output.py", line 248, in build_bytecode_output
    return f"0x{compiler_data.bytecode.hex()}"
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 150, in bytecode
    self.assembly, is_runtime=False, no_bytecode_metadata=self.no_bytecode_metadata
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 141, in assembly
    return generate_assembly(self.ir_nodes, self.no_optimize)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 126, in ir_nodes
    ir, ir_runtime, sigs = self._ir_output
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 122, in _ir_output
    return generate_ir_nodes(self.global_ctx, self.no_optimize)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 117, in global_ctx
    return GlobalContext(self.vyper_module_folded)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 107, in vyper_module_folded
    module, storage_layout = self._folded_module
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 101, in _folded_module
    return generate_folded_ast(
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 229, in generate_folded_ast
    validate_semantics(vyper_module_folded, interface_codes)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/analysis/__init__.py", line 12, in validate_semantics
    add_module_namespace(vyper_ast, interface_codes)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/analysis/module.py", line 43, in add_module_namespace
    ModuleAnalyzer(vy_module, interface_codes, namespace)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/analysis/module.py", line 77, in __init__
    self.visit(node)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/analysis/common.py", line 20, in visit
    visitor_fn(node, *args)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/analysis/module.py", line 257, in visit_EventDef
    obj = EventT.from_EventDef(node)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/types/user.py", line 235, in from_EventDef
    return cls(base_node.name, members, indexed)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/types/user.py", line 153, in __init__
    self.event_id = int(keccak256(self.signature.encode()).hex(), 16)
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/types/user.py", line 167, in signature
    return f"{self.name}({','.join(v.canonical_abi_type for v in self.arguments.values())})"
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/types/user.py", line 167, in <genexpr>
    return f"{self.name}({','.join(v.canonical_abi_type for v in self.arguments.values())})"
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/types/base.py", line 131, in canonical_abi_type
    return self.abi_type.selector_name()
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/types/base.py", line 106, in abi_type
    raise CompilerPanic("Method must be implemented by the inherited class")
vyper.exceptions.CompilerPanic: Method must be implemented by the inherited class

This is an unhandled internal compiler error. Please create an issue on Github to notify the developers.
https://github.com/vyperlang/vyper/issues/new?template=bug.md
```

```Bash
Error compiling: tests/customs/code.vy
Traceback (most recent call last):
  File "/Users/trocher/Documents/thesis/vyper/venv/bin/vyper", line 11, in <module>
    load_entry_point('vyper==0.3.8', 'console_scripts', 'vyper')()
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 57, in _parse_cli_args
    return _parse_args(sys.argv[1:])
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 154, in _parse_args
    compiled = compile_files(
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 294, in compile_files
    compiler_data = vyper.compile_codes(
  File "/Users/trocher/Documents/thesis/vyper/vyper/evm/opcodes.py", line 226, in _wrapper
    return fn(*args, **kwargs)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/__init__.py", line 141, in compile_codes
    exc_handler(contract_name, exc)
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 189, in exc_handler
    raise exception
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/__init__.py", line 138, in compile_codes
    out[contract_name][output_format] = OUTPUT_FORMATS[output_format](compiler_data)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/output.py", line 248, in build_bytecode_output
    return f"0x{compiler_data.bytecode.hex()}"
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 150, in bytecode
    self.assembly, is_runtime=False, no_bytecode_metadata=self.no_bytecode_metadata
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 141, in assembly
    return generate_assembly(self.ir_nodes, self.no_optimize)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 126, in ir_nodes
    ir, ir_runtime, sigs = self._ir_output
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 122, in _ir_output
    return generate_ir_nodes(self.global_ctx, self.no_optimize)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 258, in generate_ir_nodes
    ir_nodes, ir_runtime, function_sigs = module.generate_ir_for_module(global_ctx)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/module.py", line 165, in generate_ir_for_module
    runtime, internal_functions = _runtime_ir(runtime_functions, all_sigs, global_ctx)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/module.py", line 103, in _runtime_ir
    func_ir = generate_ir_for_function(func_ast, all_sigs, global_ctx, skip_nonpayable_check)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/function_definitions/common.py", line 62, in generate_ir_for_function
    o = generate_ir_for_external_function(code, sig, context, skip_nonpayable_check)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/function_definitions/external_function.py", line 182, in generate_ir_for_external_function
    body += [parse_body(code.body, context, ensure_terminated=True)]
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/stmt.py", line 414, in parse_body
    ir = parse_stmt(stmt, context)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/stmt.py", line 388, in parse_stmt
    return Stmt(stmt, context).ir_node
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/stmt.py", line 40, in __init__
    self.ir_node = fn()
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/stmt.py", line 71, in parse_AnnAssign
    return make_setter(lhs, rhs)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/core.py", line 819, in make_setter
    check_assign(left, right)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/core.py", line 778, in check_assign
    FAIL()
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/core.py", line 762, in FAIL
    raise TypeCheckFailure(f"assigning {right.typ} to {left.typ} {left} {right}")
vyper.exceptions.TypeCheckFailure: assigning HashMap[uint256, uint256] to HashMap[uint256, uint256] 64 0 <self.b>

This is an unhandled internal compiler error. Please create an issue on Github to notify the developers.
https://github.com/vyperlang/vyper/issues/new?template=bug.md
```

```Bash
Error compiling: tests/customs/code.vy
Traceback (most recent call last):
  File "/Users/trocher/Documents/thesis/vyper/venv/bin/vyper", line 11, in <module>
    load_entry_point('vyper==0.3.8', 'console_scripts', 'vyper')()
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 57, in _parse_cli_args
    return _parse_args(sys.argv[1:])
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 154, in _parse_args
    compiled = compile_files(
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 294, in compile_files
    compiler_data = vyper.compile_codes(
  File "/Users/trocher/Documents/thesis/vyper/vyper/evm/opcodes.py", line 226, in _wrapper
    return fn(*args, **kwargs)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/__init__.py", line 141, in compile_codes
    exc_handler(contract_name, exc)
  File "/Users/trocher/Documents/thesis/vyper/vyper/cli/vyper_compile.py", line 189, in exc_handler
    raise exception
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/__init__.py", line 138, in compile_codes
    out[contract_name][output_format] = OUTPUT_FORMATS[output_format](compiler_data)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/output.py", line 248, in build_bytecode_output
    return f"0x{compiler_data.bytecode.hex()}"
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 150, in bytecode
    self.assembly, is_runtime=False, no_bytecode_metadata=self.no_bytecode_metadata
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 141, in assembly
    return generate_assembly(self.ir_nodes, self.no_optimize)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 126, in ir_nodes
    ir, ir_runtime, sigs = self._ir_output
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/functools.py", line 966, in __get__
    val = self.func(instance)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 122, in _ir_output
    return generate_ir_nodes(self.global_ctx, self.no_optimize)
  File "/Users/trocher/Documents/thesis/vyper/vyper/compiler/phases.py", line 258, in generate_ir_nodes
    ir_nodes, ir_runtime, function_sigs = module.generate_ir_for_module(global_ctx)
  File "/Users/trocher/Documents/thesis/vyper/vyper/codegen/module.py", line 152, in generate_ir_for_module
    sig = FunctionSignature.from_definition(f, global_ctx)
  File "/Users/trocher/Documents/thesis/vyper/vyper/ast/signatures/function_signature.py", line 171, in from_definition
    assert return_type.abi_type.selector_name()
  File "/Users/trocher/Documents/thesis/vyper/vyper/semantics/types/base.py", line 106, in abi_type
    raise CompilerPanic("Method must be implemented by the inherited class")
vyper.exceptions.CompilerPanic: Method must be implemented by the inherited class

This is an unhandled internal compiler error. Please create an issue on Github to notify the developers.
https://github.com/vyperlang/vyper/issues/new?template=bug.md
```

**Comment by trocher** [comment]
Updated the issue as I found out Events are not the only place where type checking as been relaxed

**Comment by charles-cooper** [comment]
these are fixed as of 3c83947223e19a526e27bbeed69141021b82e060. the examples fail with the following errors in master (4f9f8133191c25841e309ad26f3124dbe2a46b21):

```vyper
Error compiling: repros/3232.vy
vyper.exceptions.InstantiationException: event A(uint256) is not instantiable in storage
  contract "repros/3232.vy:3", line 3:2 
       2     a:uint256
  ---> 3 x:A
  ---------^
       4
```
```vyper
Error compiling: repros/3232.vy
vyper.exceptions.InstantiationException: event A(uint256) is not instantiable in storage
  contract "repros/3232.vy:3", line 3:19 
       2     a:uint256
  ---> 3 x: HashMap[uint256,A]
  --------------------------^
       4
```
```vyper
Error compiling: repros/3232.vy
vyper.exceptions.InstantiationException: event A(uint256) is not instantiable 
  contract "repros/3232.vy:4", line 4:6 
       3 struct B:
  ---> 4     a:A
  -------------^
       5
```
```vyper
Error compiling: repros/3232.vy
vyper.exceptions.InstantiationException: HashMap[uint256, uint256] is not instantiable 
  contract "repros/3232.vy:2", line 2:6 
       1 event Foo:
  ---> 2     a:HashMap[uint256, uint256]
  -------------^
       3
```
```vyper
Error compiling: repros/3232.vy
vyper.exceptions.InstantiationException: HashMap[uint256, uint256] is not instantiable in memory
  contract "repros/3232.vy:5", function "bar", line 5:6 
       4 def bar():
  ---> 5     a:HashMap[uint256, uint256] = self.b
  -------------^
       6
```
````vyper
Error compiling: repros/3232.vy
vyper.exceptions.InstantiationException: HashMap[uint256, uint256] is not instantiable in memory
  contract "repros/3232.vy:3", function "foo", line 3:13 
       2 @external
  ---> 3 def foo() -> HashMap[uint256, uint256]:
  --------------------^
       4     return self.x
```
